### DISCUSSION TITLE
New units element?

### DISCUSSION THREAD

**Discussion by eliotwrobson** [PrairieLearn/PrairieLearn#6784] [Dev]
There was a previously proposed unit input element: https://github.com/PrairieLearn/PrairieLearn/pull/4801
It has a lot of code changes, but most of those can actually be removed by adding an external dependency (see https://pint.readthedocs.io/en/stable/) instead of having the code changes be present internally.

However, the element as described in the PR has very similar semantics to the existing `pl-number-input` element, so I'm wondering whether it's worth making this an add-on to the that existing element. What are people's thoughts on this?

Associated draft PR is here: https://github.com/PrairieLearn/PrairieLearn/pull/6790

**Comment by firasm** [comment]
As a teacher of physics courses, this would be great. We have some engineers working with us on a shared question bank on PrairieLearn that are forced to use imperial units (probably because they're still prevalent in the real world and the standards body requires questions of that sort).

Since there's no way to force a specific variant for individual courses/assessments (I think that's coming right 🤞🏽) we often have two different versions of questions. 

Sometimes we can compromise and then we build Frankenstein questions like this (one part in imperial units, and one in SI units):

<img width="848" alt="Screenshot 2022-12-20 at 9 58 11 PM" src="https://user-images.githubusercontent.com/2507459/208832557-eb4fde56-f538-49e3-ac09-3acefe127adf.png">

This way engineering students are exposed to Imperial units, and I can make sure students know how to convert from silly units to proper units.

Anyway, long story short, I'd find this useful integrated with the number-input element!

**Reply by eliotwrobson** [reply]
Do you think you would ever ask a question that specifically asked for units alone and not an associated value? That edge case I think is the main thing giving me pause on having the proposed units element just be an extension of the existing number input element.

**Reply by firasm** [reply]
Hmm. 

I don't think so, and if I did (for e.g., what are the units of Energy), I probably wouldn't default to using the number-input element anyway. I'd try the multiple choice, checkbox, or drop-down elements instead. 

But perhaps I'm not imaginative enough...

**Reply by eliotwrobson** [reply]
I think that makes sense. I might still add in the units-only input then, the only real gripe is that it seems weird to use the number input to ask for just a unit 😅 naming aside I think your use case is exactly what this feature is made for, and from discussion and the number of upvotes this seems like it would be a useful thing.

**Reply by eliotwrobson** [reply]
Looking at this again, although there are similarities with the units and number input element, I think there are actually enough semantic differences that it makes sense to introduce a whole new element. The issue is that the number input allows a bunch of extra options (like complex numbers, for example) and checking / disallowing all these different cases in functionality is going to be really tricky. In addition, I think it's going to be very hard to update the answer format of the number input element in a backwards compatible way.

This is kinda like the situation with the big-O element, where it uses similar backend to an existing element and there's _almost_ enough to combine the two, but actually doing so is actually quite hard.

**Reply by eliotwrobson** [reply]
@firasm Draft of the new element is here!: https://github.com/PrairieLearn/PrairieLearn/pull/6790

Still needs some polishing and more example questions but the core grading logic is working.

**Comment by eliotwrobson** [comment]
A question I had while implementing: Is there a use case for questions with unit agnostic grading? So for example, students could submit their answers in feet or meters and have their answer automatically converted to meters and compared to the reference solution.

**Reply by jonatanschroeder** [reply]
Yes, I'd love to have that option. I had at least one student that used the wrong unit in their answer (seconds instead of milliseconds), this would probably mitigate that issue. In that case, it would be great to have an option to specify the tolerance in a specific unit (e.g., accurate to 0.1 milliseconds, rather than number of digits or significant figures).

**Reply by eliotwrobson** [reply]
That use case raises an interesting point, the relative and absolute tolerances that can be set during grading can be scaled by the unit in this type of grading.

**Reply by jonatanschroeder** [reply]
I think relative not really, but absolute for sure. Decimal digits would also be affected by the unit. Significant figures to some extent too, but not that much.

**Reply by eliotwrobson** [reply]
Yes, absolute is really the important thing here. Seems like this element is going to have to support a few different grading modes, so I'll add some tests that help document the behavior across a few of these.

**Reply by eliotwrobson** [reply]
Related to this: Would it be useful to have the ability to restrict to a particular system of units while still being agnostic? Say, only metric units allowed for certain questions? I feel like this is a little overkill, unless someone is asking some weird problem in mixed unit systems or a student converts an answer to a different system for no reason.

**Reply by eliotwrobson** [reply]
Draft of the element with this feature is ready!: https://github.com/PrairieLearn/PrairieLearn/pull/6790

**Comment by eliotwrobson** [comment]
@firasm Element has been deployed! https://github.com/PrairieLearn/PrairieLearn/discussions/7306

**Comment by firasm** [comment]
Thank you!! is there a news item I could share with others ?

**Comment by firasm** [comment]
Update: ah yes I see it in https://github.com/PrairieLearn/PrairieLearn/pull/7302

**Reply by eliotwrobson** [reply]
Yeah unfortunately the news item didn't get merged as part of this, but the example question is synced as part of the example course. Please feel free to share the link to the documentation!: https://prairielearn.readthedocs.io/en/latest/elements/#pl-units-input-element

**Comment by eliotwrobson** [comment]
Since the element was merged in https://github.com/PrairieLearn/PrairieLearn/pull/6790, going to go ahead and close this discussion. Can be reopened if there are issues related to this element that warrant greater back and forth.
